File: 7_day/solve.py
f = open("input", "r")
lines = f.readlines()
f.close()

right_list = []
left_list = []

for line in lines:
    left, right = line.split(":")
    left_list.append(int(left.strip()))
    right_list.append(list(map(int, right.strip().split())))

def recur1(i, j, curr):
    return curr == left_list[j] if i == len(right_list[j]) else recur1(i + 1, j, curr + right_list[j][i]) or recur1(i + 1, j, curr * right_list[j][i])

def recur2(i, j, curr):
    return curr == left_list[j] if i == len(right_list[j]) else recur2(i + 1, j, curr + right_list[j][i]) or recur2(i + 1, j, curr * right_list[j][i]) or recur2(i + 1, j, int(str(curr) + str(right_list[j][i])))

part1, part2 = 0,0
for j in range(len(left_list)):
    if recur1(1, j, right_list[j][0]):
        part1 += left_list[j]

part2 = 0
for j in range(len(left_list)):
    if recur2(1, j, right_list[j][0]):
        part2 += left_list[j]
    
print("Part1:", part1)
print("Part2:", part2)

File: 19_day/solve.cpp
#include <iostream>
#include <fstream>
#include <unordered_set>
#include <vector>
#include <sstream>
#include <cctype>
#include <unordered_map>

using namespace std;


string cleanWord(const string& word) {
    string cleaned;
    for (char ch : word) {
        if (isalnum(ch)) {
            cleaned += ch;
        }
    }
    return cleaned;
}

void readFiles(unordered_set<string>& wordSet, vector<string>& wordVector, int f) {
    string line;
    string fd = (f == 1) ? "input.txt" : "trial.txt";
    ifstream file(fd);

    if (getline(file, line)) {
        istringstream iss(line);
        string word;
        while (iss >> word) {
            wordSet.insert(cleanWord(word));
        }
    }

    while (getline(file, line)) {
        istringstream iss(line);
        string word;
        while (iss >> word) {
            wordVector.push_back(cleanWord(word));
        }
    }

    file.close();
}

class Solution {
public:
    unordered_map<int, bool> memo;
    unordered_map<int, long long> cache;
    bool recur1(const unordered_set<string>& wordSet, string& word, int pos, int n) {
        if (pos == n) return true;
        
        if (memo.find(pos) != memo.end()) return memo[pos];
        for (int i = 1; i <= (n-pos); i++) {
            string curr = word.substr(pos, i);
            if (wordSet.find(curr) != wordSet.end()) {
                if (recur1(wordSet, word, pos+i, n)) return memo[pos] = true;
            }
        }
        return memo[pos] = false;
    }
    long long recur2(const unordered_set<string>& wordSet, string& word, int pos, int n) {
        if (pos == n) return 1;
        if (cache.find(pos) != cache.end()) return cache[pos];
        
        long long result = 0;
        for (int i = 1; i <= (n-pos); i++) {
            string curr = word.substr(pos, i);
            if (wordSet.find(curr) != wordSet.end()) {
                result += recur2(wordSet, word, pos+i, n);
            }
        }
        return cache[pos] = result;
    }
};

int part_1(const unordered_set<string>& wordSet, const vector<string>& wordVector) {
    Solution solver;
    int ans = 0;
    for (string word : wordVector) {
        solver.memo.clear();
        ans += solver.recur1(wordSet, word, 0, word.size());
    }
    return ans;
}

long long part_2(const unordered_set<string>& wordSet, const vector<string>& wordVector) {
    Solution solver;
    long long ans = 0;
    for (string word : wordVector) {
        solver.cache.clear();
        ans += solver.recur2(wordSet, word, 0, word.size());
    }
    return ans;
}

int main() {
    int fd = 1, target;
    cout << "Enter the file (0 for trial | 1 for input): ";
    cin >> fd;

    unordered_set<string> wordSet;
    vector<string> wordVector;

    readFiles(wordSet, wordVector, fd);

    int part1 = part_1(wordSet, wordVector);
    cout << "Part1: " << part1 << endl;
    long long part2 = part_2(wordSet, wordVector);
    cout << "Part2: " << part2 << endl;

    return 0;
}
